libp2p/src/builder.rs

```diff
@@ -0,0 +1,1203 @@
+                construct_behaviour_builder!(
+                    self,
+                    $dnsTcp,
+                    libp2p_noise::Config::new(&self.keypair).map_err(Into::<AuthenticationError>::into)?
+                )
+            }
+        }
+    };
+}
+
+impl_websocket_noise_builder!(
+    "async-std",
+    AsyncStd,
+    libp2p_dns::DnsConfig::system(libp2p_tcp::async_io::Transport::new(
+        libp2p_tcp::Config::default(),
+    ))
+);
+// TODO: Unnecessary await for Tokio Websocket (i.e. tokio dns). Not ideal but don't know a better way.
+impl_websocket_noise_builder!(
+    "tokio",
+    Tokio,
+    futures::future::ready(libp2p_dns::TokioDnsConfig::system(
+        libp2p_tcp::tokio::Transport::new(libp2p_tcp::Config::default())
+    ))
+);
+
+#[derive(Debug, thiserror::Error)]
+pub enum WebsocketError {
+    #[error("Dns")]
+    #[cfg(any(feature = "tls", feature = "noise"))]
+    Authentication(#[from] AuthenticationError),
+    #[cfg(feature = "dns")]
+    #[error("Dns")]
+    Dns(#[from] std::io::Error),
+}
+
+pub struct BehaviourBuilder<P, R> {
+    keypair: libp2p_identity::Keypair,
+    relay_behaviour: R,
+    transport: libp2p_core::transport::Boxed<(libp2p_identity::PeerId, StreamMuxerBox)>,
+    phantom: PhantomData<P>,
+}
+
+#[cfg(feature = "relay")]
+impl<P> BehaviourBuilder<P, libp2p_relay::client::Behaviour> {
+    pub fn with_behaviour<B>(
+        self,
+        mut constructor: impl FnMut(
+            &libp2p_identity::Keypair,
+            libp2p_relay::client::Behaviour,
+        ) -> Result<B, Box<dyn std::error::Error>>,
+    ) -> Result<Builder<P, B>, Box<dyn std::error::Error>> {
+        Ok(Builder {
+            behaviour: constructor(&self.keypair, self.relay_behaviour)?,
+            keypair: self.keypair,
+            transport: self.transport,
+            phantom: PhantomData,
+        })
+    }
+}
+
+impl<P> BehaviourBuilder<P, NoRelayBehaviour> {
+    pub fn with_behaviour<B>(
+        self,
+        mut constructor: impl FnMut(&libp2p_identity::Keypair) -> Result<B, Box<dyn std::error::Error>>,
+    ) -> Result<Builder<P, B>, Box<dyn std::error::Error>> {
+        // Discard `NoRelayBehaviour`.
+        let _ = self.relay_behaviour;
+
+        Ok(Builder {
+            behaviour: constructor(&self.keypair)?,
+            keypair: self.keypair,
+            transport: self.transport,
+            phantom: PhantomData,
+        })
+    }
+}
+
+pub struct Builder<P, B> {
+    keypair: libp2p_identity::Keypair,
+    behaviour: B,
+    transport: libp2p_core::transport::Boxed<(libp2p_identity::PeerId, StreamMuxerBox)>,
+    phantom: PhantomData<P>,
+}
+
+#[cfg(feature = "async-std")]
+impl<B: libp2p_swarm::NetworkBehaviour> Builder<AsyncStd, B> {
+    pub fn build(self) -> libp2p_swarm::Swarm<B> {
+        libp2p_swarm::SwarmBuilder::with_async_std_executor(
+            self.transport,
+            self.behaviour,
+            self.keypair.public().to_peer_id(),
+        )
+        .build()
```

Comment:
Long-term, I'd imagine that we ditch the other `SwarmBuilder` and only provide a single API for constructing a `Swarm`.